stats/cloud/collector.go

```diff
@@ -196,41 +202,54 @@ func (c *Collector) Link() string {
 // at regular intervals and when the context is terminated.
 func (c *Collector) Run(ctx context.Context) {
 	wg := sync.WaitGroup{}
-
+	quit := make(chan struct{})
 	// If enabled, start periodically aggregating the collected HTTP trails
 	if c.config.AggregationPeriod.Duration > 0 {
 		wg.Add(1)
 		aggregationTicker := time.NewTicker(time.Duration(c.config.AggregationCalcInterval.Duration))
 
 		go func() {
+			defer wg.Done()
 			for {
 				select {
+				case <-c.stopSendingMetricsCh:
+					return
 				case <-aggregationTicker.C:
 					c.aggregateHTTPTrails(time.Duration(c.config.AggregationWaitPeriod.Duration))
 				case <-ctx.Done():
 					c.aggregateHTTPTrails(0)
 					c.flushHTTPTrails()
-					c.pushMetrics()
-					wg.Done()
+					close(quit)
 					return
 				}
 			}
 		}()
 	}
 
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for range c.pushMetricsCh {
+			c.pushMetrics()
+		}
+	}()
+
```

Comment:
I did not mean to create a separate goroutine to pushmetrics. I am not certain this adds anything given that now the only place where we pushmetrics is in the loop below which was and still is executed only by one goroutine ;).